### DISCUSSION TITLE
Thousands separators

### DISCUSSION THREAD

**Discussion by Erenedim** [typst/typst#5142] [Q&A] [answered]
I seem to be missing something obvious her, but how to you show numbers with thousands separators?

My variable is e.g. an integer (or might be a float):
`#let n = 123452112`

How can I show this with thousands separators in my document?

`[My number is #n]`
My number is 123,452,112

**Comment by LordBaryhobal** [comment]
I personally use my trusty formatting function. I don't know of any builtin functionality to do that though
```typst
#let fmt-number(
  number,
  decimals: 2,
  decimal-sep: ".",
  thousands-sep: "'"
) = {
  let integer = calc.trunc(number)
  let decimal = calc.fract(number)
  let res = str(integer).clusters()
                        .rev()
                        .chunks(3)
                        .map(c => c.join(""))
                        .join(thousands-sep)
                        .rev()
  
  if decimals != 0 {
    res += decimal-sep
    decimal = decimal * calc.pow(10, decimals)
    decimal = calc.round(decimal)
    decimal = str(decimal) + ("0" * decimals)
    res += decimal.slice(0, decimals)
  }

  return res
}
```

**Reply by Erenedim** [reply]
Thank you so much for sharing!

**Reply by Erenedim** [reply, marked as answer]
Just want to add I found an error in your code.

If you format the number 3.08233 or 3.008233 it will be displayed as 3.82 (removing the zero in the decimal).

I have made an updated format function redoing the decimals part of the function and renaming decimals to digits to make it interchangeable with the build-in round function:

```
#let fmt_number(
  number,
  digits: 3,
  decimal_sep: ".",
  thousands_sep: ","
) = {
  let integer = calc.trunc(number)
  let decimal = calc.fract(number)
  let res = str(integer).clusters()
                        .rev()
                        .chunks(3)
                        .map(c => c.join(""))
                        .join(thousands_sep)
                        .rev()
  if digits != 0 {
    res += decimal_sep
    decimal = calc.round(decimal, digits: digits)
    let decimal_len = str(decimal).len() - 2
    if digits > decimal_len {
      decimal = str(decimal) + ("0" * (digits - decimal_len))
    }
    decimal = str(decimal).slice(2)
    res += decimal
  }
  return res
}
```

**Comment by Mc-Zen** [comment]
Hi @Erenedim ,

this also sounds like a use-case for [zero](https://typst.app/universe/package/zero), a package dedicated to number formatting. You can show numbers of all kinds via the central `num()` function. The package provides support for grouping (e.g., into thousands), truncating/rounding, scientific notation and treats several edge cases. Moreover, there are configurable options. For example, the default separator is a thin space but you can configure it via the command `set-group()`. 

```typ
#import "@preview/zero:0.2.0": num, set-group
#set-group(separator: ",")

#let n = 123452112
#num(n)
```
![image](https://github.com/user-attachments/assets/5456ca9e-7df9-46c6-9881-1ebf85ef14fd)

**Reply by Erenedim** [reply]
Thanks for your input, I will have a look at this package also :)
